File: ruby/difference-of-squares/difference_of_squares.rb
# frozen_string_literal: true

# Calculate square of sum and sum of squares
class Squares
  def initialize(num)
    @nums = (1..num)
  end

  def square_of_sum
    @_square_of_sum ||= calc_square_of_sum
    @_square_of_sum
  end

  def sum_of_squares
    @_sum_of_squares ||= calc_sum_of_squares
    @_sum_of_squares
  end

  def difference
    @_difference ||= calc_difference
    @_difference
  end

  private

  def calc_square_of_sum
    @nums.sum**2
  end

  def calc_sum_of_squares
    @nums.map { |x| x**2 }.sum
  end

  def calc_difference
    square_of_sum - sum_of_squares
  end
end

File: javascript/difference-of-squares/difference-of-squares.js
export class Squares {
  constructor(n) {
    this._sumOfSquares = calcSumOfSquares(n);
    this._squareOfSum = calcSquareOfSum(n);
    this._difference = this._squareOfSum - this._sumOfSquares;
  }

  get sumOfSquares() {
    return this._sumOfSquares;
  }

  get squareOfSum() {
    return this._squareOfSum;
  }

  get difference() {
    return this._difference;
  }
}

const calcSumOfSquares = (n) => {
  return n * (n + 1) * (2 * n + 1) / 6;
};

const calcSquareOfSum = (n) => {
  let sum = n * (n + 1) / 2;
  return sum * sum;
};
